File: compilers/java/src/test/java/runar/compiler/passes/C17NotNotGuardTest.java
package runar.compiler.passes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import org.junit.jupiter.api.Test;
import runar.compiler.Cli;
import runar.compiler.frontend.ParserDispatch;
import runar.compiler.ir.anf.AnfProgram;
import runar.compiler.ir.ast.ContractNode;
import runar.compiler.ir.stack.OpcodeOp;
import runar.compiler.ir.stack.PushOp;
import runar.compiler.ir.stack.PushValue;
import runar.compiler.ir.stack.RollOp;
import runar.compiler.ir.stack.StackOp;
import runar.compiler.ir.stack.StackProgram;

/**
 * Regression test for deep-review finding C17 (unsound {@code not-not-elim}).
 *
 * <p>{@code OP_NOT OP_NOT} is boolean NORMALISATION, not numeric identity: for a
 * non-canonical operand (say 5) the pair yields 1, while deleting it leaves 5.
 * Truthiness survives, the VALUE does not — and a downstream {@code OP_EQUAL} /
 * {@code OP_NUMEQUAL} consumes the value, so the optimised and unoptimised
 * programs disagree on accept/reject.
 *
 * <p>The old rule was an unguarded 2-op window. Composed with the sibling
 * {@code PUSH 0; OP_NUMEQUAL → OP_NOT} rewrite — which manufactures a fresh
 * {@code OP_NOT} sitting on an ARBITRARY script number — it collapsed
 * {@code x !== 0n} all the way down to {@code x}: the fold-OFF script
 * {@code 007c7c9c91517c7c9c} optimised to {@code 519c} (compare the witness
 * against 1 instead of against "x is non-zero"), rejecting the valid spend
 * {@code x = 5}.
 *
 * <p>The rule is now a 3-op window that includes the PRODUCER of the negated
 * value and only fires when that producer provably leaves a canonical 0/1.
 */
class C17NotNotGuardTest {

    /** The boolean is consumed by value (=== true), so the collapse is observable. */
    private static final String SRC =
        "import { SmartContract, assert } from 'runar-lang';\n"
        + "class C17Check extends SmartContract {\n"
        + "  constructor() { super(); }\n"
        + "  public check(x: bigint): void { assert((x !== 0n) === true); }\n"
        + "}\n";

    private static StackProgram lower(String src, String file) throws Exception {
        ContractNode contract = ParserDispatch.parse(src, file);
        Validate.run(contract);
        contract = ExpandFixedArrays.run(contract);
        Typecheck.run(contract);
        AnfProgram anf = AnfLower.run(contract);
        anf = Cli.optimizeAnf(anf, true); // fold-OFF, matching the conformance goldens
        return StackLower.run(anf);
    }

    /**
     * End-to-end: with the peephole ON the emitted script must still compute
     * "x is non-zero" rather than degenerating into "x". The unguarded rule
     * produced {@code 519c}; the guarded one keeps the {@code OP_NOT OP_NOT}
     * normalisation as {@code 9191519c}.
     */
    @Test
    void notEqualsZeroSurvivesThePeephole() throws Exception {
        StackProgram raw = lower(SRC, "C17Check.runar.ts");
        String foldOffPeepholeOff = Emit.run(raw);
        String foldOffPeepholeOn = Emit.run(Peephole.run(raw));

        assertEquals("007c7c9c91517c7c9c", foldOffPeepholeOff,
            "baseline (peephole OFF) must be PUSH0 SWAP SWAP NUMEQUAL NOT PUSH1 SWAP SWAP NUMEQUAL");
        assertEquals("9191519c", foldOffPeepholeOn,
            "peephole ON must keep the OP_NOT OP_NOT boolean normalisation (C17)");
    }

    /** The same statement, at the Stack-IR level: the NOT pair must survive. */
    @Test
    void peepholeKeepsTheNotPairInLoweredOps() throws Exception {
        List<StackOp> ops = Peephole.optimize(lower(SRC, "C17Check.runar.ts").methods().get(0).ops());
        long nots = ops.stream().filter(o -> o instanceof OpcodeOp c && "OP_NOT".equals(c.code())).count();
        assertEquals(2, nots, "both OP_NOTs must remain — deleting them changes the compared VALUE");
    }

    /**
     * Guard: the producer is a stack shuffle, so its provenance is invisible to
     * this local window. The pair must survive.
     */
    @Test
    void nonCanonicalProducerKeepsTheNotPair() {
        List<StackOp> r = Peephole.optimize(List.of(
            new RollOp(3), new OpcodeOp("OP_NOT"), new OpcodeOp("OP_NOT")));
        assertEquals(3, r.size(), "OP_NOT OP_NOT over a rolled (arbitrary) value must not be deleted");
        assertTrue(r.get(1) instanceof OpcodeOp a && "OP_NOT".equals(a.code()));
        assertTrue(r.get(2) instanceof OpcodeOp b && "OP_NOT".equals(b.code()));
    }

    /** Guard: a bare pair with no visible producer is equally unprovable. */
    @Test
    void bareNotPairWithNoVisibleProducerSurvives() {
        List<StackOp> r = Peephole.optimize(List.of(new OpcodeOp("OP_NOT"), new OpcodeOp("OP_NOT")));
        assertEquals(2, r.size());
    }

    /**
     * NEGATIVE case — the rule must not become dead. A provably canonical
     * producer still gets its redundant normalisation removed.
     */
    @Test
    void canonicalProducerStillCollapses() {
        for (String producer : new String[] {
            "OP_EQUAL", "OP_NUMEQUAL", "OP_NUMNOTEQUAL", "OP_LESSTHAN", "OP_GREATERTHAN",
            "OP_LESSTHANOREQUAL", "OP_GREATERTHANOREQUAL", "OP_BOOLAND", "OP_BOOLOR",
            "OP_WITHIN", "OP_NOT", "OP_0NOTEQUAL", "OP_CHECKSIG", "OP_CHECKMULTISIG"
        }) {
            List<StackOp> r = Peephole.optimize(List.of(
                new OpcodeOp(producer), new OpcodeOp("OP_NOT"), new OpcodeOp("OP_NOT")));
            assertEquals(1, r.size(), producer + " leaves a canonical bool — the NOT pair is redundant");
            assertTrue(r.get(0) instanceof OpcodeOp o && producer.equals(o.code()));
        }
    }

    /** Literal boolean / 0 / 1 pushes are canonical producers too. */
    @Test
    void canonicalLiteralPushesStillCollapse() {
        for (PushValue v : new PushValue[] {
            PushValue.of(true), PushValue.of(false), PushValue.of(0), PushValue.of(1)
        }) {
            List<StackOp> r = Peephole.optimize(List.of(
                new PushOp(v), new OpcodeOp("OP_NOT"), new OpcodeOp("OP_NOT")));
            assertEquals(1, r.size(), v + " is already canonical");
            assertTrue(r.get(0) instanceof PushOp);
        }
    }

    /** A non-canonical literal push must keep the pair (PUSH 5 NOT NOT != PUSH 5). */
    @Test
    void nonCanonicalLiteralPushKeepsTheNotPair() {
        List<StackOp> r = Peephole.optimize(List.of(
            new PushOp(PushValue.of(5)), new OpcodeOp("OP_NOT"), new OpcodeOp("OP_NOT")));
        assertEquals(3, r.size(), "PUSH 5 OP_NOT OP_NOT yields 1, not 5");
    }

    /**
     * The exact composition that made this a bug: the sibling
     * {@code PUSH 0; OP_NUMEQUAL → OP_NOT} rewrite must not hand the unguarded
     * rule a second OP_NOT to eat.
     */
    @Test
    void push0NumequalNotChainKeepsBothNots() {
        List<StackOp> r = Peephole.optimize(List.of(
            new RollOp(3),
            new PushOp(PushValue.of(0)),
            new OpcodeOp("OP_NUMEQUAL"),
            new OpcodeOp("OP_NOT")));
        assertEquals(3, r.size(), "expected Roll(3) OP_NOT OP_NOT, got " + r);
        assertTrue(r.get(0) instanceof RollOp);
        assertTrue(r.get(1) instanceof OpcodeOp a && "OP_NOT".equals(a.code()));
        assertTrue(r.get(2) instanceof OpcodeOp b && "OP_NOT".equals(b.code()));
    }
}

File: compilers/go/codegen/c17_not_not_guard_test.go
package codegen

import (
	"fmt"
	"strings"
	"testing"
)

// ---------------------------------------------------------------------------
// C17 — `OP_NOT OP_NOT` is boolean NORMALISATION, not numeric identity.
//
// For a non-canonical operand (say 5) the pair yields 1, while deleting it
// leaves 5. Truthiness is preserved, the VALUE is not — and a downstream
// OP_EQUAL / OP_NUMEQUAL / state serialisation consumes the value, so the
// optimised and unoptimised programs disagree on accept/reject.
//
// The unguarded 2-op `not-not-elim` rule is unsound in composition with the
// sibling `PUSH 0; OP_NUMEQUAL -> OP_NOT` rule, which synthesises a fresh
// OP_NOT sitting on an ARBITRARY script number. The stack lowerer emits
// `!==` as [OP_NUMEQUAL, OP_NOT] (see stack.go binaryOpMap), so
//
//	x !== 0n   ==>   <x> ; PUSH 0 ; OP_NUMEQUAL ; OP_NOT
//
// becomes `<x> ; OP_NOT ; OP_NOT`, which the unguarded rule then deleted
// entirely — `x !== 0n` compiled to `x`.
//
// The fix widens the rule to a 3-op window that includes the PRODUCER of the
// negated value and only fires when that producer provably leaves a canonical
// boolean (0 or 1) behind.
// ---------------------------------------------------------------------------

// c17Signature renders a stack-op list as a compact, comparable string.
func c17Signature(ops []StackOp) string {
	parts := make([]string, len(ops))
	for i, op := range ops {
		switch op.Op {
		case "opcode":
			parts[i] = op.Code
		case "push":
			switch op.Value.Kind {
			case "bigint":
				parts[i] = fmt.Sprintf("PUSH(%s)", op.Value.BigInt.String())
			case "bool":
				parts[i] = fmt.Sprintf("PUSH(%t)", op.Value.Bool)
			default:
				parts[i] = fmt.Sprintf("PUSH(%x)", op.Value.Bytes)
			}
		default:
			parts[i] = strings.ToUpper(op.Op)
		}
	}
	return strings.Join(parts, " ")
}

func c17Check(t *testing.T, name string, in []StackOp, want string) {
	t.Helper()
	got := c17Signature(OptimizeStackOps(in))
	if got != want {
		t.Errorf("%s:\n  input:    %s\n  got:      %s\n  expected: %s",
			name, c17Signature(in), got, want)
	}
}

func c17PushBool(b bool) StackOp {
	return StackOp{Op: "push", Value: PushValue{Kind: "bool", Bool: b}}
}

// The whole point of C17: when the producer of the negated value is NOT
// provably a canonical boolean, the OP_NOT pair must SURVIVE.
func TestC17_NotNotElim_NonCanonicalProducer_PairSurvives(t *testing.T) {
	// `x !== 0n` where x comes off the witness stack via a pick.
	c17Check(t, "pick-produced operand (x !== 0n)",
		[]StackOp{
			{Op: "pick", Depth: 3},
			pushBigIntOp(0),
			opcodeOp("OP_NUMEQUAL"),
			opcodeOp("OP_NOT"),
		},
		"PICK OP_NOT OP_NOT")

	// Mirrored composition: `(!b) === false` lowers to OP_NOT; PUSH 0; OP_NUMEQUAL.
	c17Check(t, "mirrored composition ((!b) === false)",
		[]StackOp{
			{Op: "pick", Depth: 2},
			opcodeOp("OP_NOT"),
			pushBigIntOp(0),
			opcodeOp("OP_NUMEQUAL"),
		},
		"PICK OP_NOT OP_NOT")

	// Arbitrary numeric producers.
	c17Check(t, "OP_ADD producer",
		[]StackOp{opcodeOp("OP_ADD"), opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
		"OP_ADD OP_NOT OP_NOT")
	c17Check(t, "OP_SIZE producer",
		[]StackOp{opcodeOp("OP_SIZE"), opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
		"OP_SIZE OP_NOT OP_NOT")
	c17Check(t, "non-canonical literal producer",
		[]StackOp{pushBigIntOp(5), opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
		"PUSH(5) OP_NOT OP_NOT")

	// Stack-shuffling ops are DELIBERATELY excluded from the canonical set:
	// they forward a value whose provenance this local window cannot see.
	c17Check(t, "dup producer (provenance invisible)",
		[]StackOp{{Op: "dup"}, opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
		"DUP OP_NOT OP_NOT")
	c17Check(t, "roll producer (provenance invisible)",
		[]StackOp{{Op: "roll", Depth: 4}, opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
		"ROLL OP_NOT OP_NOT")

	// A bare pair with no visible producer must also survive — the operand
	// could be anything at all.
	c17Check(t, "bare pair, no producer in window",
		[]StackOp{opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
		"OP_NOT OP_NOT")
}

// NEGATIVE case: the guard must not make the rule dead. A genuinely canonical
// producer still gets the redundant pair removed.
func TestC17_NotNotElim_CanonicalProducer_StillElided(t *testing.T) {
	canonical := []string{
		"OP_EQUAL",
		"OP_NUMEQUAL",
		"OP_NUMNOTEQUAL",
		"OP_LESSTHAN",
		"OP_GREATERTHAN",
		"OP_LESSTHANOREQUAL",
		"OP_GREATERTHANOREQUAL",
		"OP_BOOLAND",
		"OP_BOOLOR",
		"OP_WITHIN",
		"OP_0NOTEQUAL",
		"OP_CHECKSIG",
		"OP_CHECKMULTISIG",
	}
	for _, code := range canonical {
		c17Check(t, code+" producer",
			[]StackOp{opcodeOp(code), opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
			code)
	}

	// OP_NOT itself is canonical: NOT NOT NOT collapses to a single NOT.
	c17Check(t, "OP_NOT producer (triple negation)",
		[]StackOp{opcodeOp("OP_NOT"), opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
		"OP_NOT")

	// Literal canonical booleans.
	c17Check(t, "PUSH 0 producer",
		[]StackOp{pushBigIntOp(0), opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
		"PUSH(0)")
	c17Check(t, "PUSH 1 producer",
		[]StackOp{pushBigIntOp(1), opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
		"PUSH(1)")
	c17Check(t, "PUSH true producer",
		[]StackOp{c17PushBool(true), opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
		"PUSH(true)")
	c17Check(t, "PUSH false producer",
		[]StackOp{c17PushBool(false), opcodeOp("OP_NOT"), opcodeOp("OP_NOT")},
		"PUSH(false)")
}

// The raw_bytes peephole barrier must keep holding for the widened window.
func TestC17_NotNotElim_RawBytesBarrier(t *testing.T) {
	c17Check(t, "raw_bytes inside the window",
		[]StackOp{
			opcodeOp("OP_NUMEQUAL"),
			{Op: "raw_bytes", RawBytes: []byte{0x51}, InArity: 0, OutArity: 1},
			opcodeOp("OP_NOT"),
			opcodeOp("OP_NOT"),
		},
		"OP_NUMEQUAL RAW_BYTES OP_NOT OP_NOT")
}
